src/software/thunderscope/gl/layers/gl_cost_vis_layer.py

```diff
@@ -10,6 +12,39 @@
 from software.thunderscope.thread_safe_buffer import ThreadSafeBuffer
 from software.thunderscope.gl.layers.gl_layer import GLLayer
 from software.thunderscope.gl.graphics.gl_heatmap import GLHeatmap
+from software.thunderscope.gl.graphics.gl_gradient_legend import GLGradientLegend
+
+from software.thunderscope.constants import Colors, DepthValues
+
+
+class GLCostVisOverlayLayer(GLLayer):
+    """Overlay for GLCostVisLayer"""
+
+    def __init__(self, gradient: QtGui.QLinearGradient) -> None:
+        """Initialize the GLCostVisOverlayLayer
+        
+        :param gradient: The gradient to use for coloring the legend
+
+        """
+
+        super().__init__("GLCostVisOverlayLayer")
+        self.setDepthValue(DepthValues.OVERLAY_DEPTH)
+
+        self.gradient = gradient
+        self.legend_graphic: GLGradientLegend = None
+
+    def refresh_graphics(self) -> None:
+        """Update graphics in this layer"""
+
+        if not self.legend_graphic:
+            self.legend_graphic = GLGradientLegend(
+                parent_item=self,
+                size=(10, 100),
+                offset=(16, -16),
+                gradient=self.gradient,
+                labels={"1.0": 1, "0.5": 0.5, "0.0": 0},
```

Comment:
I think the legend would be improved if it wasnt static and it was the actual range of the combined cost values of the currently drawn cost functions. E.g. when we show two cost functions and the max cost ends up being 0.1, the legend should show up to 0.1. This way, the viewer can have some reference of the magnitude of the actual cost values and how they change as we add/remove cost functions.

Reply:
Ah I see what you mean now. I didn't realize the cost functions were additive

Reply:
Yes, exactly. The `ratePass` function multiplies the results of a set of cost functions (each returning a cost between 0 and 1): 

https://github.com/UBC-Thunderbots/Software/blob/de49d34541f1a4659b539289ff394db75e6e5feb/src/software/ai/passing/cost_function.cpp#L40-L41

It's worth noting that the current implementation only supports visualizing 4 of the 6 cost values.

https://github.com/UBC-Thunderbots/Software/blob/de49d34541f1a4659b539289ff394db75e6e5feb/src/software/ai/passing/cost_function.cpp#L407-L408

Missing `in_region_quality` and `pass_speed_quality`. `pass_speed_quality` is a bit difficult to add since we don't know the pass speed (unless we assume it by adding the visualization parameters list). But we might be able to add `in_region_quality` as well. That value should actually be constant for visualization as well, so no need to recalculate it every tick.

Reply:
Okay nice. Just to double check though, this would mean that the gradient shown in the legend doesn't necessirally match the colors on the field? Since the colors on the field are normalized between 0 and 0.8.

Reply:
Before I was scaling the data points from the range [0, 1] down to [0, 0.8] to determine the color, but now I'm using ColorMap to map the values to colors. I've also given the heatmap some transparency. The data isn't really normalized (i.e. the max data point doesn't become white); there's a linear relation between the range of scalar values and the color scale

Reply:
So does this mean that if the max cost is 0.1, the square that is associated with that cost will look almost black?

Reply:
Yes, this is also how it was implemented in the previous cost visualizer